pkg/volume/flexvolume/flexvolume_util.go

```diff
@@ -0,0 +1,142 @@
+)
+
+// FlexVolumeDriverStatus represents the return value of the driver callout.
+type FlexVolumeDriverStatus struct {
+	// Status of the callout. One of "Success" or "Failure".
+	Status string
+	// Message is the reason for failure.
+	Message string
+	/* Device assigned by the driver. */
+	Device string `json:"device"`
+}
+
+// flexVolumeUtil is the utility structure to setup and teardown devices from
+// the host.
+type flexVolumeUtil struct{}
+
+func (u *flexVolumeUtil) init(plugin *flexVolumePlugin) error {
+	// call the init script
+	output, err := exec.New().Command(plugin.getExecutable(), initCmd).CombinedOutput()
+	if err != nil {
+		glog.Errorf("Failed to init driver: %s, error: %s", plugin.driverName, err.Error())
+		return err
+	}
+	var status FlexVolumeDriverStatus
+	if err := json.Unmarshal(output, &status); err != nil {
+		glog.Errorf("Failed to unmarshal output from driver: %s, output: %s, error: %s", plugin.driverName, output, err.Error())
+		return err
+	} else if status.Status != StatusSuccess {
+		errMsg := fmt.Sprintf("Failed to init driver: %s, status: %s, error: %s", plugin.driverName, status.Status, status.Message)
+		glog.Errorf(errMsg)
+		return fmt.Errorf("%s", errMsg)
+	}
+
+	glog.V(5).Infof("Successfully initialized driver %s", plugin.driverName)
+	return nil
+}
+
+// Attach exposes a volume on the host.
+func (u *flexVolumeUtil) setUp(f *flexVolumeBuilder, dir string) (string, error) {
+	execPath := f.execPath
+
+	var options string
+	if f.options != nil {
+		out, err := json.Marshal(f.options)
+		if err != nil {
+			glog.Errorf("Failed to marshal plugin options, error: %s", err.Error())
+			return "", err
+		}
+		if len(out) != 0 {
+			options = string(out)
+		} else {
+			options = ""
+		}
+	}
+
+	// Executable provider commamd.
+	cmd := f.runner.Command(execPath, setUpCmd, f.podNamespace, f.volName, f.podName, string(f.podUID), options)
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		glog.Errorf("Failed to setup the volume %s, output: %s, error: %s", f.volName, output, err.Error())
+		return "", err
+	}
+
+	var status FlexVolumeDriverStatus
+	if err := json.Unmarshal(output, &status); err != nil {
+		glog.Errorf("Failed to unmarshal output from driver for volume: %s, output: %s, error: %s", f.volName, output, err.Error())
+		return "", err
+	} else if status.Status != StatusSuccess {
+		errMsg := fmt.Sprintf("Failed to setup volume: %s, status: %s, error: %s", f.volName, status.Status, status.Message)
+		glog.Errorf(errMsg)
+		return "", fmt.Errorf("%s", errMsg)
+	}
+
+	glog.Infof("Successfully setup volume %s on device: %s", f.volName, status.Device)
+
+	return status.Device, nil
+}
+
+// Detach teardown a volume from the host.
+func (u *flexVolumeUtil) teardown(f *flexVolumeCleaner, dir, mntDevice string) (string, error) {
+	execPath := f.execPath
+
+	// Executable provider command.
+	cmd := f.runner.Command(execPath, tearDownCmd, f.podNamespace, f.podName, string(f.podUID), f.volName, mntDevice)
```

Comment:
As mentioned before since the setup and teardown scripts should respectively only be responsible for attaching and detaching a volume, they have no need be aware of the pod namespace, pod name, or pod UID.

I really prefer to keep interfaces as minimal as possible. In this case, I don't want plugin writers to do anything with that information. So I strongly suggest we remove these from the script interface (don't pass them in, and update the documentation/examples accordingly).